pkg/operator/clustermembercontroller2/clustermembercontroller.go

```diff
@@ -0,0 +1,382 @@
+	return nil
+}
+
+func (c *ClusterMemberController) getEtcdClient() (*clientv3.Client, error) {
+	endpoints, err := c.Endpoints()
+	if err != nil {
+		return nil, err
+	}
+
+	dialOptions := []grpc.DialOption{
+		grpc.WithBlock(), // block until the underlying connection is up
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      etcdCertFile,
+		KeyFile:       etcdKeyFile,
+		TrustedCAFile: etcdTrustedCAFile,
+	}
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	cfg := &clientv3.Config{
+		DialOptions: dialOptions,
+		Endpoints:   endpoints,
+		DialTimeout: 5 * time.Second,
+		TLS:         tlsConfig,
+	}
+
+	cli, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cli, err
+}
+
+func (c *ClusterMemberController) Endpoints() ([]string, error) {
+	hostEtcd, err := c.endpointsLister.Endpoints(clustermembercontroller.EtcdEndpointNamespace).Get(clustermembercontroller.EtcdEndpointName)
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorGettingHostEtcd", "error occured while getting host-etcd endpoint: %#v", err)
+		return []string{}, err
+	}
+	if len(hostEtcd.Subsets) == 0 {
+		c.eventRecorder.Warningf("EtcdAddressNotFound", "could not find etcd address in host-etcd")
+		return []string{}, fmt.Errorf("could not find etcd address in host-etcd")
+	}
+	var endpoints []string
+	for _, addr := range hostEtcd.Subsets[0].Addresses {
+		endpoints = append(endpoints, fmt.Sprintf("https://%s.%s:2379", addr.Hostname, c.etcdDiscoveryDomain))
+	}
+	return endpoints, nil
+}
+
+func (c *ClusterMemberController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(workQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(workQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(workQueueKey) },
+	}
+}
+
+func (c *ClusterMemberController) isEtcdHealthy() (bool, error) {
+	// getting a new client everytime because we dont know what etcd-membership looks like
+	etcdClient, err := c.getEtcdClient()
+	defer etcdClient.Close()
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorGettingEtcdClient", "error getting etcd client: %#v", err)
+		return false, err
+	}
+
+	memberList, err := etcdClient.MemberList(context.Background())
+	if err != nil {
+		c.eventRecorder.Warningf("ErrorGettingMemberList", "error getting etcd member list: %#v", err)
+		return false, err
+	}
+	for _, member := range memberList.Members {
+		statusResp, err := etcdClient.Status(context.Background(), member.ClientURLs[0])
+		if err != nil {
+			c.eventRecorder.Warningf("EtcdMemberNotHealthy", "etcd member %s is not healthy: %#v", member.Name, err)
+			// since error is indicative of unhealthy member, not returning
+			// the actual error
+			return false, nil
+		}
+		klog.V(4).Infof("etcd member %s is healthy committed and with %s index", member.Name, statusResp.RaftIndex)
+	}
+	return true, nil
+}
+
+func (c *ClusterMemberController) getPeerToScale() (string, []error) {
```

Comment:
this signature is wrong.  You can end up with many unreadyEtcdPeers.  Rename the method, fix the signature

Reply:
reworking